core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -0,0 +1,395 @@
+        }
+
+        return redirectCtx.originalUri().equals(redirectUri) &&
+               redirectCtx.request().method() == newHeaders.method();
+    }
+
+    private static String buildUri(ClientRequestContext ctx, RequestHeaders headers) {
+        final String originalUri;
+        try (TemporaryThreadLocals threadLocals = TemporaryThreadLocals.acquire()) {
+            final StringBuilder sb = threadLocals.stringBuilder();
+            if (SessionProtocol.isHttp(ctx.sessionProtocol())) {
+                sb.append(SessionProtocol.HTTP.uriText());
+            } else {
+                sb.append(SessionProtocol.HTTPS.uriText());
+            }
+            sb.append("://");
+            String authority = headers.authority();
+            final Endpoint endpoint = ctx.endpoint();
+            assert endpoint != null;
+            if (authority == null) {
+                authority = endpoint.authority();
+            }
+            setAuthorityAndPort(ctx, endpoint, sb, authority);
+            sb.append(headers.path());
+            originalUri = sb.toString();
+        }
+        return originalUri;
+    }
+
+    private static void setAuthorityAndPort(ClientRequestContext ctx, Endpoint endpoint, StringBuilder sb,
+                                            String authority) {
+        // Add port number as well so that we don't raise a CyclicRedirectsException when the port is
+        // different.
+
+        if (authority.charAt(0) == '[') {
+            final int closingBracketPos = authority.lastIndexOf(']');
+            if (closingBracketPos < 0) {
+                // Should never reach here because we already validate the authority.
+                throw new IllegalStateException("Invalid authority: " + authority);
+            }
+            sb.append(authority);
+            if (authority.indexOf(':', closingBracketPos) < 0) {
+                // The authority does not have port number so we add it.
+                addPort(ctx, endpoint, sb);
+            }
+            return;
+        }
+
+        if (NetUtil.isValidIpV6Address(authority)) {
+            sb.append('[');
+            sb.append(authority);
+            sb.append(']');
+            addPort(ctx, endpoint, sb);
+            return;
+        }
+
+        sb.append(authority);
+        if (authority.lastIndexOf(':') < 0) {
+            addPort(ctx, endpoint, sb);
+        }
+    }
+
+    private static void addPort(ClientRequestContext ctx, Endpoint endpoint, StringBuilder sb) {
+        sb.append(':');
+        sb.append(endpoint.port(ctx.sessionProtocol().defaultPort()));
+    }
+
+    @VisibleForTesting
+    static class RedirectContext {
+
+        private final HttpRequest request;
+        private final CompletableFuture<Void> responseWhenComplete;
+        private final CompletableFuture<HttpResponse> responseFuture;
+        private final Multimap<HttpMethod, String> redirectUris = LinkedListMultimap.create();
+        private final String originalUri;
```

Comment:
How about lazily creating these. If there is no redirection, the fields might not be accessed.

Reply:
That's a good suggestion. 👍